src/ruby/puppetserver-lib/puppet/server/master.rb

```diff
@@ -127,6 +162,79 @@ def getTaskData(environment_name, module_name, task_name)
 
   private
 
+  # @return Puppet::Node::Facts facts, Hash trusted_facts
+  def process_facts(request_data)
+    if request_data['facts'].nil?
+      facts = get_facts_from_pdb(request_data['certname'], request_data['environment'])
+    else
+      facts_from_request = request_data['facts']
```

Comment:
If facts from request does not contain the key `values` (or that key is anything other than a Hash) then the Facts object will be invalid and blow up later.

Reply:
Is that something we'd want to validate here, or at the schema level in Clojure as we're parsing the body?

Reply:
It'd be nice to have any user input, if given, be syntactically correct. (though I kinda hate surfacing schema errors to users) I'm also not totally sure what other values from [`Node.from_data_hash`](https://github.com/puppetlabs/puppet/blob/master/lib/puppet/node/facts.rb#L43-L59) are required, if any.

Reply:
Non-answer of the day

Reply:
I am now checking more cases when processing facts, making sure that we won't blow up from things being nil at least.